File: Java/One Edit Distance.java
/*
Given two strings S and T, determine if they are both one edit distance apart. Hint:
1. If | n – m | is greater than 1, we know immediately both are not one-edit distance apart.
2. It might help if you consider these cases separately, m == n and m ≠ n.
3. Assume that m is always ≤ n, which greatly simplifies the conditional statements.
If m > n, we could just simply swap S and T.
4. If m == n, it becomes finding if there is exactly one modified operation. 
If m != n, you do not have to consider the delete operation. Just consider the insert operation in T.
*/

public boolean isOneEditDistance(String s, String t) {
        int m = s.length(), n = t.length();
        if (m > n) return isOneEditDistance(t, s);
        if (n - m > 1) return false;
        int i = 0, shift = n - m;
        while (i < m && s.charAt(i) == t.charAt(i)) i++;
        //s == t, which returns false
        //or n = m+1 and last character of t is redundant
        if (i == m) return shift>0;
        if (shift == 0) i++;
        while (i < m && s.charAt(i) == t.charAt(i + shift)) i++;
        return i == m;
    }

File: Python/One Edit Distance.py
"""
Given two strings S and T, determine if they are both one edit distance apart.
"""

class Solution(object):
    def isOneEditDistance(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        if abs(len(s) - len(t)) > 1: return False
        i, j, flag = 0, 0, False
        
        while i < len(s) and j < len(t):
            if s[i] == t[j]:
                i += 1
                j += 1
            elif s[i] != t[j] and flag:
                return False
            else:
                if len(s) == len(t):
                    flag = True
                    i += 1
                    j += 1
                elif len(s) < len(t):
                    flag = True
                    j += 1
                else:
                    flag = True
                    i += 1
        if i < len(s) or j < len(t) and flag is False:
            return True
        if i == len(s) and j == len(t) and flag:
            return True
        return False

class Solution(object):
    def isOneEditDistance(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        m, n = len(s), len(t)
        if m > n:
            return self.isOneEditDistance(t, s)
        if n - m > 1:
            return False
        
        i, shift = 0, n - m
        while i < m and s[i] == t[i]:
            i += 1
        if shift == 0:
            i += 1
        while i < m and s[i] == t[i + shift]:
            i += 1
            
        return i == m
